app/src/org/commcare/models/database/user/UserDbUpgradeUtils.java

```diff
@@ -198,4 +196,54 @@ protected static void migrateFormRecordsToV3(Context c, SQLiteDatabase db) {
             newStorage.write(newRecord);
         }
     }
+
+    protected static void migrateV4FormRecords(Context c, SQLiteDatabase db) {
+        SqlStorage<FormRecordV4> oldStorage = getFormRecordStorage(c, db, FormRecordV4.class);
+
+        // Transform old records to new add to newRecords
+        Vector<FormRecord> newRecords = new Vector<>();
+        for (FormRecordV4 oldRecord : oldStorage) {
+            FormRecord newRecord = new FormRecord(
+                    oldRecord.getStatus(),
+                    oldRecord.getFormNamespace(),
+                    oldRecord.getAesKey(),
+                    oldRecord.getInstanceID(),
+                    oldRecord.lastModified(),
+                    oldRecord.getAppId());
+
+            // Merge Record's instance
+            String instanceUri = oldRecord.getInstanceURIString();
+            Cursor cursor = c.getContentResolver().query(Uri.parse(instanceUri), null, null, null, null);
+            if (cursor != null && cursor.getCount() > 0) {
```

Comment:
quick note that technically cursor operations should be wrapped in try{}closes

Reply:
We don't need to catch any exceptions, we just need a try{}finally{} block where the cursor gets closed.

Reply:
Gotcha!